Sources/MockoloFramework/Operations/ProtocolMapGenerator.swift

```diff
@@ -16,94 +16,134 @@
+            fatalError("Annotation is invalid: \(annotation)")
+        }
+        if let queue = queue {
+            let lock = NSLock()
+            
+            scanPaths(dirs) { filePath in
+                _ = semaphore?.wait(timeout: DispatchTime.distantFuture)
+                queue.async {
+                    generateProtcolMap(filePath,
+                                       exclusionSuffixes: exclusionSuffixes,
+                                       annotationData: annotationData,
+                                       lock: lock,
+                                       process: process)
+                    semaphore?.signal()
+                }
+            }
+            
+            // Wait for queue to drain
+            queue.sync(flags: .barrier) {}
+        } else {
+            scanPaths(dirs) { filePath in
                 generateProtcolMap(filePath,
                                    exclusionSuffixes: exclusionSuffixes,
-                                   annotation: annotation,
-                                   lock: lock,
+                                   annotationData: annotationData,
+                                   lock: nil,
                                    process: process)
-                semaphore?.signal()
             }
         }
-        
-        // Wait for queue to drain
-        queue.sync(flags: .barrier) {}
-    } else {
+    default:
+        var treeVisitor = EntityVisitor(annotation: annotation)
         scanPaths(dirs) { filePath in
             generateProtcolMap(filePath,
                                exclusionSuffixes: exclusionSuffixes,
                                annotation: annotation,
+                               treeVisitor: &treeVisitor,
                                lock: nil,
                                process: process)
         }
     }
 }
 
-
 private func generateProtcolMap(files: [String],
                                 exclusionSuffixes: [String]? = nil,
-                                annotation: Data,
+                                annotation: String,
+                                parserType: ParserType,
                                 semaphore: DispatchSemaphore?,
                                 queue: DispatchQueue?,
                                 process: @escaping ([Entity]) -> ()) {
-    if let queue = queue {
-        let lock = NSLock()
-        for filePath in files {
-            _ = semaphore?.wait(timeout: DispatchTime.distantFuture)
-            queue.async {
+    
+    switch parserType {
+    case .sourceKit:
+        guard let annotationData = annotation.data(using: .utf8) else {
+            fatalError("Annotation is invalid: \(annotation)")
+        }
+        
+        if let queue = queue {
+            let lock = NSLock()
+            for filePath in files {
+                _ = semaphore?.wait(timeout: DispatchTime.distantFuture)
+                queue.async {
```

Comment:
Seems like batch async processing is a common pattern in Mockolo, there should be a utility function for this similar to dispatch_function in GCD.

Reply:
I agree, though in the 3 copies, there are 2 distinct uses. Two where we have an array and it's a lot like the GCD one and then one where the scanPath is scanning directories slowly and producing filenames one at a time and we dispatch them as they become available one by one. I suppose the make the API for the latter case and it would still work in the array case.

Reply:
Will address this on a separate pr as this is not related to swiftsyntax skeleton code.